consensus/state_sync/state_sync.go

```diff
@@ -0,0 +1,102 @@
+)
+
+type StateSyncModule interface {
+	modules.Module
+
+	// -- Constructor Setter Functions --
+
+	// `HandleStateSync` function:
+	// - Create a Utility Context
+	// - Block.ValidateBasic()
+	// - Consensus Module Replica Path
+	//   - Prepare Block:  utilityContext.SetProposalBlock(block)
+	//   - Apply Block:    utilityContext.ApplyBlock(block)
+	//   - Validate Block: utilityContext.AppHash == Block.AppHash
+	//   - Store Block:    consensusModule.CommitBlock()
+	HandleStateSyncMessage(msg BlockResponseMessage)
+
+	// `GetPeerSyncMeta` function:
+	// - Retrieve a list of active peers with their metadata (identified and retrieved through P2P's `Churn Management`)
+	GetPeerMetadata(GetPeerSyncMeta func() (peers []PeerSyncMeta, err error))
+
+	// `NetworkSend` function contract:
+	// - sends data to an address via P2P network
+	NetworkSend(NetworkSend func(data []byte, address cryptoPocket.Address) error)
+
+	// -- Sync modes --
+
+	// In the StateSync protocol, the Node fields valid BlockRequests from its peers to help them CatchUp to be Synced.
+	// This sub-protocol is continuous throughout the lifecycle of StateSync.
+	RunServerMode()
+
+	// In SyncedMode, the Node is caught up to the latest block and is listening & waiting for the latest block to be passed
+	// to maintain a synchronous state with the global SyncState.
+	// - UpdatePeerMetadata from P2P module
+	// - UpdateSyncState
+	// - Rely on new blocks to be propagated via the P2P network after Validators reach consensus
+	// - If `localSyncState.Height < globalNetworkSyncState.Height` -> RunSyncMode() // careful about race-conditions
+	RunSyncedMode()
+
+	// Runs sync mode 'service' that continuously runs while `localSyncState.Height < globalNetworkSyncState.Height`
+	// - UpdatePeerMetadata from P2P module
+	// - Retrieve missing blocks from peers
+	// - Process retrieved blocks
+	// - UpdateSyncState
+	// - If `localSyncState.Height == globalNetworkSyncState.Height` -> RunSyncedMode()
+	RunSyncMode()
+
+	// Returns the `highest priority aka lowest height` missing block heights up to `max` heights
+	GetMissingBlockHeights(state SyncState, max int) (blockHeights []int64, err error)
+
+	// Random selection of eligilbe peers enables a fair distribution of blockRequests over time via law of large numbers
+	// An eligible peer is when `PeerMeta.MinHeight <= blockHeight <= PeerMeta.MaxHeight`
+	GetRandomEligiblePeersForHeight(blockHeight int64) (eligiblePeer PeerSyncMeta, err error)
+
+	// Uses `NetworkSend` to send a `BlockRequestMessage` to a specific peer
+	SendBlockRequest(peerId string) error
+
+	// Uses 'NetworkSend' to send a `BlockResponseMessage` to a specific peer
+	// This function is used in 'ServerMode()'
+	HandleBlockRequest(message BlockRequestMessage) error
+
+	// Uses `HandleBlock` to process retrieved blocks from peers
+	// Must update sync state using `SetMissingBlockHeight`
+	ProcessBlock(block *typesCons.Block) error
+}
+
+type SyncState interface {
```

Comment:
In the `.md` file, we define `PeerSyncAggregate` and `PeerSyncMeta`, but neither is used here.

Are the former structs just for illustrative purposes (not implementation)?

Reply:
> GetPeers() []PeerSyncMeta

> type PeerSyncMeta interface {

It is used a few times. I think it lines up